packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,572 @@
+  final AutocompleteFieldBuilder fieldBuilder;
+
+  /// Builds the selectable options widgets from a list of options objects.
+  final AutocompleteOptionsBuilder<T> optionsBuilder;
+
+  /// Returns the string to display in the field when the option is selected.
+  ///
+  /// This is useful when using a custom T type and the string to display is
+  /// different than the string to search by.
+  ///
+  /// If not provided, will use `option.toString()`.
+  final AutocompleteOptionToString<T> displayStringForOption;
+
+  /// Called when an option is selected by the user.
+  final AutocompleteOnSelected<T>? onSelected;
+
+  /// A function that returns the current selectable options objects given the
+  /// current TextEditingValue.
+  final AutocompleteBuildOptions<T> buildOptions;
+
+  // The default way to convert an option to a string.
+  static String _defaultStringForOption<T>(T option) {
+    return option.toString();
+  }
+
+  @override
+  _AutocompleteCoreState<T> createState() => _AutocompleteCoreState<T>();
+}
+
+class _AutocompleteCoreState<T> extends State<AutocompleteCore<T>> {
+  final GlobalKey _fieldKey = GlobalKey();
+  final LayerLink _optionsLayerLink = LayerLink();
+  final TextEditingController _textEditingController = TextEditingController();
+  List<T> _options = <T>[];
+  T? _selection;
+
+  // The OverlayEntry containing the options.
+  OverlayEntry? _floatingOptions;
+
+  // True iff the state indicates that the options should be visible.
+  bool get _shouldShowOptions {
+    final TextSelection selection = _textEditingController.selection;
+    final bool fieldIsFocused = selection.baseOffset >= 0
+        && selection.extentOffset >= 0;
+    final bool hasOptions = _options != null && _options.isNotEmpty;
+    return fieldIsFocused && _selection == null && hasOptions;
+  }
+
+  // Called when _textEditingController changes.
+  void _onChangedField() {
+    final List<T> options = widget.buildOptions(
+      _textEditingController.value,
+    );
+    assert(options != null);
+    setState(() {
+      _options = options;
+      if (_selection != null) {
+        final String selectionString = widget.displayStringForOption(_selection!);
+        if (_textEditingController.text == selectionString) {
+          return;
+        }
+      }
+      _selection = null;
+    });
+    _updateOverlay();
+  }
+
+  // Called from fieldBuilder when the user submits the field.
+  void _onFieldSubmitted() {
+    if (_options.isEmpty) {
+      return;
+    }
+    _select(_options[0]);
+  }
+
+  // Select the given option and update the widget.
+  void _select(T? nextSelection) {
+    if (nextSelection == _selection) {
+      return;
+    }
+    setState(() {
+      _selection = nextSelection;
+      final String selectionString = nextSelection == null
+          ? ''
+          : widget.displayStringForOption(nextSelection);
+      _textEditingController.value = TextEditingValue(
```

Comment:
This is not going to trigger input formatters which is fine. But what about the `EditableText.onChanged` callback (this should be treated as user input right)?  I guess the developer can still figure out what kind of change since there's an `onSelected` callback. But it is a bit annoying. Maybe we should move more text editing responsibilities to `TextEditingController`.

Reply:
Right, I double checked and it won't call `onChanged` when the user selects something.  I'm not sure what users would expect.  I'll make a note about it in the documentation for `onSelected` at least.